// mongodb/lib/mongodb/gridfs/chunk.js
var Binary = require('bson').Binary,
  ObjectID = require('bson').ObjectID;

/**
 * Class for representing a single chunk in GridFS.
 *
 * @class
 *
 * @param file {GridStore***REMOVED*** The {@link GridStore***REMOVED*** object holding this chunk.
 * @param mongoObject {object***REMOVED*** The mongo object representation of this chunk.
 *
 * @throws Error when the type of data field for {@link mongoObject***REMOVED*** is not
 *     supported. Currently supported types for data field are instances of
 *     {@link String***REMOVED***, {@link Array***REMOVED***, {@link Binary***REMOVED*** and {@link Binary***REMOVED***
 *     from the bson module
 *
 * @see Chunk#buildMongoObject
 */
var Chunk = exports.Chunk = function(file, mongoObject, writeConcern) {
  if(!(this instanceof Chunk)) return new Chunk(file, mongoObject);

  this.file = file;
  var self = this;
  var mongoObjectFinal = mongoObject == null ? {***REMOVED*** : mongoObject;
  this.writeConcern = writeConcern || {w:1***REMOVED***;
  this.objectId = mongoObjectFinal._id == null ? new ObjectID() : mongoObjectFinal._id;
  this.chunkNumber = mongoObjectFinal.n == null ? 0 : mongoObjectFinal.n;
  this.data = new Binary();

  if(mongoObjectFinal.data == null) {
  ***REMOVED*** else if(typeof mongoObjectFinal.data == "string") {
    var buffer = new Buffer(mongoObjectFinal.data.length);
    buffer.write(mongoObjectFinal.data, 'binary', 0);
    this.data = new Binary(buffer);
  ***REMOVED*** else if(Array.isArray(mongoObjectFinal.data)) {
    var buffer = new Buffer(mongoObjectFinal.data.length);
    buffer.write(mongoObjectFinal.data.join(''), 'binary', 0);
    this.data = new Binary(buffer);
  ***REMOVED*** else if(mongoObjectFinal.data instanceof Binary || Object.prototype.toString.call(mongoObjectFinal.data) == "[object Binary]") {
    this.data = mongoObjectFinal.data;
  ***REMOVED*** else if(Buffer.isBuffer(mongoObjectFinal.data)) {
  ***REMOVED*** else {
    throw Error("Illegal chunk format");
  ***REMOVED***
  // Update position
  this.internalPosition = 0;
***REMOVED***;

/**
 * Writes a data to this object and advance the read/write head.
 *
 * @param data {string***REMOVED*** the data to write 
 * @param callback {function(*, GridStore)***REMOVED*** This will be called after executing
 *     this method. The first parameter will contain null and the second one
 *     will contain a reference to this object.
 */
Chunk.prototype.write = function(data, callback) {
  this.data.write(data, this.internalPosition);
  this.internalPosition = this.data.length();
  if(callback != null) return callback(null, this);
  return this;
***REMOVED***;

/**
 * Reads data and advances the read/write head.
 *
 * @param length {number***REMOVED*** The length of data to read.
 *
 * @return {string***REMOVED*** The data read if the given length will not exceed the end of
 *     the chunk. Returns an empty String otherwise.
 */
Chunk.prototype.read = function(length) {
  // Default to full read if no index defined
  length = length == null || length == 0 ? this.length() : length;

  if(this.length() - this.internalPosition + 1 >= length) {
    var data = this.data.read(this.internalPosition, length);
    this.internalPosition = this.internalPosition + length;
    return data;
  ***REMOVED*** else {
    return '';
  ***REMOVED***
***REMOVED***;

Chunk.prototype.readSlice = function(length) {
  if ((this.length() - this.internalPosition) >= length) {
    var data = null;
    if (this.data.buffer != null) { //Pure BSON
      data = this.data.buffer.slice(this.internalPosition, this.internalPosition + length);
    ***REMOVED*** else { //Native BSON
      data = new Buffer(length);
      length = this.data.readInto(data, this.internalPosition);
    ***REMOVED***
    this.internalPosition = this.internalPosition + length;
    return data;
  ***REMOVED*** else {
    return null;
  ***REMOVED***
***REMOVED***;

/**
 * Checks if the read/write head is at the end.
 *
 * @return {boolean***REMOVED*** Whether the read/write head has reached the end of this
 *     chunk.
 */
Chunk.prototype.eof = function() {
  return this.internalPosition == this.length() ? true : false;
***REMOVED***;

/**
 * Reads one character from the data of this chunk and advances the read/write
 * head.
 *
 * @return {string***REMOVED*** a single character data read if the the read/write head is
 *     not at the end of the chunk. Returns an empty String otherwise.
 */
Chunk.prototype.getc = function() {
  return this.read(1);
***REMOVED***;

/**
 * Clears the contents of the data in this chunk and resets the read/write head
 * to the initial position.
 */
Chunk.prototype.rewind = function() {
  this.internalPosition = 0;
  this.data = new Binary();
***REMOVED***;

/**
 * Saves this chunk to the database. Also overwrites existing entries having the
 * same id as this chunk.
 *
 * @param callback {function(*, GridStore)***REMOVED*** This will be called after executing
 *     this method. The first parameter will contain null and the second one
 *     will contain a reference to this object.
 */
Chunk.prototype.save = function(callback) {
  var self = this;

  self.file.chunkCollection(function(err, collection) {
    if(err) return callback(err);

    collection.remove({'_id':self.objectId***REMOVED***, self.writeConcern, function(err, result) {
      if(err) return callback(err);

      if(self.data.length() > 0) {
        self.buildMongoObject(function(mongoObject) {
          var options = {forceServerObjectId:true***REMOVED***;
          for(var name in self.writeConcern) {
            options[name] = self.writeConcern[name];
          ***REMOVED***

          collection.insert(mongoObject, options, function(err, collection) {
            callback(err, self);
          ***REMOVED***);
        ***REMOVED***);
      ***REMOVED*** else {
        callback(null, self);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * Creates a mongoDB object representation of this chunk.
 *
 * @param callback {function(Object)***REMOVED*** This will be called after executing this 
 *     method. The object will be passed to the first parameter and will have
 *     the structure:
 *        
 *        <pre><code>
 *        {
 *          '_id' : , // {number***REMOVED*** id for this chunk
 *          'files_id' : , // {number***REMOVED*** foreign key to the file collection
 *          'n' : , // {number***REMOVED*** chunk number
 *          'data' : , // {bson#Binary***REMOVED*** the chunk data itself
 *        ***REMOVED***
 *        </code></pre>
 *
 * @see <a href="http://www.mongodb.org/display/DOCS/GridFS+Specification#GridFSSpecification-{{chunks***REMOVED******REMOVED***">MongoDB GridFS Chunk Object Structure</a>
 */
Chunk.prototype.buildMongoObject = function(callback) {
  var mongoObject = {
    'files_id': this.file.fileId,
    'n': this.chunkNumber,
    'data': this.data***REMOVED***;
  // If we are saving using a specific ObjectId
  if(this.objectId != null) mongoObject._id = this.objectId;

  callback(mongoObject);
***REMOVED***;

/**
 * @return {number***REMOVED*** the length of the data
 */
Chunk.prototype.length = function() {
  return this.data.length();
***REMOVED***;

/**
 * The position of the read/write head
 * @name position
 * @lends Chunk#
 * @field
 */
Object.defineProperty(Chunk.prototype, "position", { enumerable: true
  , get: function () {
      return this.internalPosition;
    ***REMOVED***
  , set: function(value) {
      this.internalPosition = value;
    ***REMOVED***
***REMOVED***);

/**
 * The default chunk size
 * @constant
 */
Chunk.DEFAULT_CHUNK_SIZE = 1024 * 256;
