// mongoose/lib/namedscope.js
var Query = require('./query');
function NamedScope () {***REMOVED***

NamedScope.prototype.query;

NamedScope.prototype.where = function () {
  var q = this.query || (this.query = new Query());
  q.where.apply(q, arguments);
  return q;
***REMOVED***;

/**
 * Decorate
 *
 * @param {NamedScope***REMOVED*** target
 * @param {Object***REMOVED*** getters
 * @api private
 */

NamedScope.prototype.decorate = function (target, getters) {
  var name = this.name
    , block = this.block
    , query = this.query;
  if (block) {
    if (block.length === 0) {
      Object.defineProperty(target, name, {
        get: getters.block0(block)
      ***REMOVED***);
    ***REMOVED*** else {
      target[name] = getters.blockN(block);
    ***REMOVED***
  ***REMOVED*** else {
    Object.defineProperty(target, name, {
      get: getters.basic(query)
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

NamedScope.prototype.compile = function (model) {
  var allScopes = this.scopesByName
    , scope;
  for (var k in allScopes) {
    scope = allScopes[k];
    scope.decorate(model, {
      block0: function (block) {
        return function () {
          var cquery = this._cumulativeQuery || (this._cumulativeQuery = new Query().bind(this));
          block.call(cquery);
          return this;
        ***REMOVED***;
      ***REMOVED***,
      blockN: function (block) {
        return function () {
          var cquery = this._cumulativeQuery || (this._cumulativeQuery = new Query().bind(this));
          block.apply(cquery, arguments);
          return this;
        ***REMOVED***;
      ***REMOVED***,
      basic: function (query) {
        return function () {
          var cquery = this._cumulativeQuery || (this._cumulativeQuery = new Query().bind(this));
          cquery.find(query);
          return this;
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

module.exports = NamedScope;
