// delayed-stream/lib/delayed_stream.js
var Stream = require('stream').Stream;
var util = require('util');

module.exports = DelayedStream;
function DelayedStream() {
  this.source = null;
  this.dataSize = 0;
  this.maxDataSize = 1024 * 1024;
  this.pauseStream = true;

  this._maxDataSizeExceeded = false;
  this._released = false;
  this._bufferedEvents = [];
***REMOVED***
util.inherits(DelayedStream, Stream);

DelayedStream.create = function(source, options) {
  var delayedStream = new this();

  options = options || {***REMOVED***;
  for (var option in options) {
    delayedStream[option] = options[option];
  ***REMOVED***

  delayedStream.source = source;

  var realEmit = source.emit;
  source.emit = function() {
    delayedStream._handleEmit(arguments);
    return realEmit.apply(source, arguments);
  ***REMOVED***;

  source.on('error', function() {***REMOVED***);
  if (delayedStream.pauseStream) {
    source.pause();
  ***REMOVED***

  return delayedStream;
***REMOVED***;

DelayedStream.prototype.__defineGetter__('readable', function() {
  return this.source.readable;
***REMOVED***);

DelayedStream.prototype.resume = function() {
  if (!this._released) {
    this.release();
  ***REMOVED***

  this.source.resume();
***REMOVED***;

DelayedStream.prototype.pause = function() {
  this.source.pause();
***REMOVED***;

DelayedStream.prototype.release = function() {
  this._released = true;

  this._bufferedEvents.forEach(function(args) {
    this.emit.apply(this, args);
  ***REMOVED***.bind(this));
  this._bufferedEvents = [];
***REMOVED***;

DelayedStream.prototype.pipe = function() {
  var r = Stream.prototype.pipe.apply(this, arguments);
  this.resume();
  return r;
***REMOVED***;

DelayedStream.prototype._handleEmit = function(args) {
  if (this._released) {
    this.emit.apply(this, args);
    return;
  ***REMOVED***

  if (args[0] === 'data') {
    this.dataSize += args[1].length;
    this._checkIfMaxDataSizeExceeded();
  ***REMOVED***

  this._bufferedEvents.push(args);
***REMOVED***;

DelayedStream.prototype._checkIfMaxDataSizeExceeded = function() {
  if (this._maxDataSizeExceeded) {
    return;
  ***REMOVED***

  if (this.dataSize <= this.maxDataSize) {
    return;
  ***REMOVED***

  this._maxDataSizeExceeded = true;
  var message =
    'DelayedStream#maxDataSize of ' + this.maxDataSize + ' bytes exceeded.'
  this.emit('error', new Error(message));
***REMOVED***;
