// cors/lib/index.js
/*jslint indent: 2*/
/*global require: true, module: true*/

(function () {

  'use strict';

  var defaults = {
    origin: '*',
    methods: 'GET,HEAD,PUT,PATCH,POST,DELETE'
  ***REMOVED***;

  function configureOrigin(options, req) {
    var origin = options.origin;
    if (origin === true) {
      origin = req.headers.origin;
    ***REMOVED*** else if (!origin) {
      origin = '*';
    ***REMOVED***
    return {
      key: 'Access-Control-Allow-Origin',
      value: origin
    ***REMOVED***;
  ***REMOVED***

  function configureMethods(options) {
    var methods = options.methods || defaults.methods;
    if (methods.join) {
      methods = options.methods.join(','); // .methods is an array, so turn it into a string
    ***REMOVED***
    return {
      key: 'Access-Control-Allow-Methods',
      value: methods
    ***REMOVED***;
  ***REMOVED***

  function configureCredentials(options) {
    if (options.credentials === true) {
      return {
        key: 'Access-Control-Allow-Credentials',
        value: 'true'
      ***REMOVED***;
    ***REMOVED***
    return null;
  ***REMOVED***

  function configureAllowedHeaders(options, req) {
    var headers = options.allowedHeaders || options.headers;
    if (!headers) {
      headers = req.headers['access-control-request-headers']; // .headers wasn't specified, so reflect the request headers
    ***REMOVED*** else if (headers.join) {
      headers = headers.join(','); // .headers is an array, so turn it into a string
    ***REMOVED***
    if (headers && headers.length) {
      return {
        key: 'Access-Control-Allow-Headers',
        value: headers
      ***REMOVED***;
    ***REMOVED***
    return null;
  ***REMOVED***

  function configureExposedHeaders(options, req) {
    var headers = options.exposedHeaders;
    if (!headers) {
      return null;
    ***REMOVED*** else if (headers.join) {
      headers = headers.join(','); // .headers is an array, so turn it into a string
    ***REMOVED***
    if (headers && headers.length) {
      return {
        key: 'Access-Control-Expose-Headers',
        value: headers
      ***REMOVED***;
    ***REMOVED***
    return null;
  ***REMOVED***

  function configureMaxAge(options) {
    var maxAge = options.maxAge && options.maxAge.toString();
    if (maxAge && maxAge.length) {
      return {
        key: 'Access-Control-Max-Age',
        value: maxAge
      ***REMOVED***;
    ***REMOVED***
    return null;
  ***REMOVED***

  function cors(options, req, res, next) {
    var headers = [],
      method = req.method && req.method.toUpperCase && req.method.toUpperCase(),
      applyHeaders = function (headers, res) {
        headers.forEach(function (header) {
          if (header && header.value) {
            if (res.set) {
              res.set(header.key, header.value);
            ***REMOVED*** else {
              // for Express <4
              res.setHeader(header.key, header.value);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***;

    if (method === 'OPTIONS') {
      // preflight
      headers.push(configureOrigin(options, req));
      headers.push(configureCredentials(options, req));
      headers.push(configureMethods(options, req));
      headers.push(configureAllowedHeaders(options, req));
      headers.push(configureMaxAge(options, req));
      applyHeaders(headers, res);
      res.statusCode = 204;
      res.end();
    ***REMOVED*** else {
      // actual response
      headers.push(configureOrigin(options, req));
      headers.push(configureCredentials(options, req));
      headers.push(configureExposedHeaders(options, req));
      applyHeaders(headers, res);
      next();
    ***REMOVED***
  ***REMOVED***

  function middlewareWrapper(o) {
    // if no options were passed in, use the defaults
    if (!o) {
      o = {***REMOVED***;
    ***REMOVED***
    if (o.origin === undefined) {
      o.origin = defaults.origin;
    ***REMOVED***
    if (o.methods === undefined) {
      o.methods = defaults.methods;
    ***REMOVED***

    // if options are static (either via defaults or custom options passed in), wrap in a function
    var optionsCallback = null;
    if (typeof o === 'function') {
      optionsCallback = o;
    ***REMOVED*** else {
      /*jslint unparam: true*/ // `req` is part of the signature, but isn't used for this stub
      optionsCallback = function (req, cb) {
        cb(null, o);
      ***REMOVED***;
      /*jslint unparam: false*/
    ***REMOVED***

    return function (req, res, next) {
      optionsCallback(req, function (err, options) {
        if (err) {
          next(err);
        ***REMOVED*** else {
          var originCallback = null;
          if (options.origin && typeof options.origin === 'function') {
            originCallback = options.origin;
          ***REMOVED*** else if (options.origin) {
            /*jslint unparam: true*/ // `origin` is part of the signature, but isn't used for this stub
            originCallback = function (origin, cb) {
              cb(null, options.origin);
            ***REMOVED***;
            /*jslint unparam: false*/
          ***REMOVED***

          if (originCallback) {
            originCallback(req.headers.origin, function (err, origin) {
              if (err || !origin) {
                next(err);
              ***REMOVED*** else {
                var corsOptions = Object.create(options);
                corsOptions.origin = origin;
                cors(corsOptions, req, res, next);
              ***REMOVED***
            ***REMOVED***);
          ***REMOVED*** else {
            next();
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***;
  ***REMOVED***

  // can pass either an options hash, an options delegate, or nothing
  module.exports = middlewareWrapper;

***REMOVED***());
