// passport/lib/passport/index.js
/**
 * Module dependencies.
 */
var fs = require('fs')
  , path = require('path')
  , util = require('util')
  , Strategy = require('./strategy')
  , SessionStrategy = require('./strategies/session')
  , initialize = require('./middleware/initialize')
  , authenticate = require('./middleware/authenticate');


/**
 * `Passport` constructor.
 *
 * @api public
 */
function Passport() {
  this._key = 'passport';
  this._strategies = {***REMOVED***;
  this._serializers = [];
  this._deserializers = [];
  this._infoTransformers = [];
  this._framework = null;
  
  this._userProperty = 'user';
  
  this.use(new SessionStrategy());
***REMOVED***;

/**
 * Utilize the given `strategy` with optional `name`, overridding the strategy's
 * default name.
 *
 * Examples:
 *
 *     passport.use(new TwitterStrategy(...));
 *
 *     passport.use('api', new http.BasicStrategy(...));
 *
 * @param {String|Strategy***REMOVED*** name
 * @param {Strategy***REMOVED*** strategy
 * @return {Passport***REMOVED*** for chaining
 * @api public
 */
Passport.prototype.use = function(name, strategy) {
  if (!strategy) {
    strategy = name;
    name = strategy.name;
  ***REMOVED***
  if (!name) throw new Error('authentication strategies must have a name');
  
  this._strategies[name] = strategy;
  return this;
***REMOVED***;

/**
 * Un-utilize the `strategy` with given `name`.
 *
 * In typical applications, the necessary authentication strategies are static,
 * configured once and always available.  As such, there is often no need to
 * invoke this function.
 *
 * However, in certain situations, applications may need dynamically configure
 * and de-configure authentication strategies.  The `use()`/`unuse()`
 * combination satisfies these scenarios.
 *
 * Examples:
 *
 *     passport.unuse('legacy-api');
 *
 * @param {String***REMOVED*** name
 * @return {Passport***REMOVED*** for chaining
 * @api public
 */
Passport.prototype.unuse = function(name) {
  delete this._strategies[name];
  return this;
***REMOVED***

/**
 * Setup Passport to be used under framework.
 *
 * By default, Passport exposes middleware that operate using Connect-style
 * middleware using a `fn(req, res, next)` signature.  Other popular frameworks
 * have different expectations, and this function allows Passport to be adapted
 * to operate within such environments.
 *
 * If you are using a Connect-compatible framework, including Express, there is
 * no need to invoke this function.
 *
 * Examples:
 *
 *     passport.framework(require('hapi-passport')());
 *
 * @param {Object***REMOVED*** name
 * @return {Passport***REMOVED*** for chaining
 * @api public
 */
Passport.prototype.framework = function(fw) {
  this._framework = fw;
  return this;
***REMOVED***

/**
 * Passport's primary initialization middleware.
 *
 * This middleware must be in use by the Connect/Express application for
 * Passport to operate.
 *
 * Options:
 *   - `userProperty`  Property to set on `req` upon login, defaults to _user_
 *
 * Examples:
 *
 *     app.configure(function() {
 *       app.use(passport.initialize());
 *     ***REMOVED***);
 *
 *     app.configure(function() {
 *       app.use(passport.initialize({ userProperty: 'currentUser' ***REMOVED***));
 *     ***REMOVED***);
 *
 * @param {Object***REMOVED*** options
 * @return {Function***REMOVED*** middleware
 * @api public
 */
Passport.prototype.initialize = function(options) {
  options = options || {***REMOVED***;
  this._userProperty = options.userProperty || 'user';
  
  if (this._framework && this._framework.initialize) {
    return this._framework.initialize().bind(this);
  ***REMOVED***
  
  return initialize().bind(this);
***REMOVED***

/**
 * Middleware that will restore login state from a session.
 *
 * Web applications typically use sessions to maintain login state between
 * requests.  For example, a user will authenticate by entering credentials into
 * a form which is submitted to the server.  If the credentials are valid, a
 * login session is established by setting a cookie containing a session
 * identifier in the user's web browser.  The web browser will send this cookie
 * in subsequent requests to the server, allowing a session to be maintained.
 *
 * If sessions are being utilized, and a login session has been established,
 * this middleware will populate `req.user` with the current user.
 *
 * Note that sessions are not strictly required for Passport to operate.
 * However, as a general rule, most web applications will make use of sessions.
 * An exception to this rule would be an API server, which expects each HTTP
 * request to provide credentials in an Authorization header.
 *
 * Examples:
 *
 *     app.configure(function() {
 *       app.use(connect.cookieParser());
 *       app.use(connect.session({ secret: 'keyboard cat' ***REMOVED***));
 *       app.use(passport.initialize());
 *       app.use(passport.session());
 *     ***REMOVED***);
 *
 * Options:
 *   - `pauseStream`      Pause the request stream before deserializing the user
 *                        object from the session.  Defaults to _false_.  Should
 *                        be set to true in cases where middleware consuming the
 *                        request body is configured after passport and the
 *                        deserializeUser method is asynchronous.
 *
 * @param {Object***REMOVED*** options
 * @return {Function***REMOVED*** middleware
 * @api public
 */
Passport.prototype.session = function(options) {
  return this.authenticate('session', options);
***REMOVED***

/**
 * Middleware that will authenticate a request using the given `strategy` name,
 * with optional `options` and `callback`.
 *
 * Examples:
 *
 *     passport.authenticate('local', { successRedirect: '/', failureRedirect: '/login' ***REMOVED***)(req, res);
 *
 *     passport.authenticate('local', function(err, user) {
 *       if (!user) { return res.redirect('/login'); ***REMOVED***
 *       res.end('Authenticated!');
 *     ***REMOVED***)(req, res);
 *
 *     passport.authenticate('basic', { session: false ***REMOVED***)(req, res);
 *
 *     app.get('/auth/twitter', passport.authenticate('twitter'), function(req, res) {
 *       // request will be redirected to Twitter
 *     ***REMOVED***);
 *     app.get('/auth/twitter/callback', passport.authenticate('twitter'), function(req, res) {
 *       res.json(req.user);
 *     ***REMOVED***);
 *
 * @param {String***REMOVED*** strategy
 * @param {Object***REMOVED*** options
 * @param {Function***REMOVED*** callback
 * @return {Function***REMOVED*** middleware
 * @api public
 */
Passport.prototype.authenticate = function(strategy, options, callback) {
  if (this._framework && this._framework.authenticate) {
    return this._framework.authenticate(strategy, options, callback).bind(this);
  ***REMOVED***
  
  return authenticate(strategy, options, callback).bind(this);
***REMOVED***

/**
 * Middleware that will authorize a third-party account using the given
 * `strategy` name, with optional `options`.
 *
 * If authorization is successful, the result provided by the strategy's verify
 * callback will be assigned to `req.account`.  The existing login session and
 * `req.user` will be unaffected.
 *
 * This function is particularly useful when connecting third-party accounts
 * to the local account of a user that is currently authenticated.
 *
 * Examples:
 *
 *    passport.authorize('twitter-authz', { failureRedirect: '/account' ***REMOVED***);
 *
 * @param {String***REMOVED*** strategy
 * @param {Object***REMOVED*** options
 * @return {Function***REMOVED*** middleware
 * @api public
 */
Passport.prototype.authorize = function(strategy, options, callback) {
  var fwAuthorize = this._framework && (this._framework.authorize || this._framework.authenticate);

  options = options || {***REMOVED***;
  options.assignProperty = 'account';

  if (fwAuthorize) {
    return fwAuthorize(strategy, options, callback).bind(this);
  ***REMOVED***
  
  return authenticate(strategy, options, callback).bind(this);
***REMOVED***

/**
 * Registers a function used to serialize user objects into the session.
 *
 * Examples:
 *
 *     passport.serializeUser(function(user, done) {
 *       done(null, user.id);
 *     ***REMOVED***);
 *
 * @api public
 */
Passport.prototype.serializeUser = function(fn, done) {
  if (typeof fn === 'function') {
    return this._serializers.push(fn);
  ***REMOVED***
  
  // private implementation that traverses the chain of serializers, attempting
  // to serialize a user
  var user = fn;
  
  var stack = this._serializers;
  (function pass(i, err, obj) {
    // serializers use 'pass' as an error to skip processing
    if ('pass' === err) {
      err = undefined;
    ***REMOVED***
    // an error or serialized object was obtained, done
    if (err || obj || obj === 0) { return done(err, obj); ***REMOVED***
    
    var layer = stack[i];
    if (!layer) {
      return done(new Error('failed to serialize user into session'));
    ***REMOVED***
    
    try {
      layer(user, function(e, o) { pass(i + 1, e, o); ***REMOVED*** )
    ***REMOVED*** catch(e) {
      return done(e);
    ***REMOVED***
  ***REMOVED***)(0);
***REMOVED***

/**
 * Registers a function used to deserialize user objects out of the session.
 *
 * Examples:
 *
 *     passport.deserializeUser(function(id, done) {
 *       User.findById(id, function (err, user) {
 *         done(err, user);
 *       ***REMOVED***);
 *     ***REMOVED***);
 *
 * @api public
 */
Passport.prototype.deserializeUser = function(fn, done) {
  if (typeof fn === 'function') {
    return this._deserializers.push(fn);
  ***REMOVED***
  
  // private implementation that traverses the chain of deserializers,
  // attempting to deserialize a user
  var obj = fn;
  
  var stack = this._deserializers;
  (function pass(i, err, user) {
    // deserializers use 'pass' as an error to skip processing
    if ('pass' === err) {
      err = undefined;
    ***REMOVED***
    // an error or deserialized user was obtained, done
    if (err || user) { return done(err, user); ***REMOVED***
    // a valid user existed when establishing the session, but that user has
    // since been removed
    if (user === null || user === false) { return done(null, false); ***REMOVED***
    
    var layer = stack[i];
    if (!layer) {
      return done(new Error('failed to deserialize user out of session'));
    ***REMOVED***
    
    try {
      layer(obj, function(e, u) { pass(i + 1, e, u); ***REMOVED*** )
    ***REMOVED*** catch(e) {
      return done(e);
    ***REMOVED***
  ***REMOVED***)(0);
***REMOVED***

/**
 * Registers a function used to transform auth info.
 *
 * In some circumstances authorization details are contained in authentication
 * credentials or loaded as part of verification.
 *
 * For example, when using bearer tokens for API authentication, the tokens may
 * encode (either directly or indirectly in a database), details such as scope
 * of access or the client to which the token was issued.
 *
 * Such authorization details should be enforced separately from authentication.
 * Because Passport deals only with the latter, this is the responsiblity of
 * middleware or routes further along the chain.  However, it is not optimal to
 * decode the same data or execute the same database query later.  To avoid
 * this, Passport accepts optional `info` along with the authenticated `user`
 * in a strategy's `success()` action.  This info is set at `req.authInfo`,
 * where said later middlware or routes can access it.
 *
 * Optionally, applications can register transforms to proccess this info,
 * which take effect prior to `req.authInfo` being set.  This is useful, for
 * example, when the info contains a client ID.  The transform can load the
 * client from the database and include the instance in the transformed info,
 * allowing the full set of client properties to be convieniently accessed.
 *
 * If no transforms are registered, `info` supplied by the strategy will be left
 * unmodified.
 *
 * Examples:
 *
 *     passport.transformAuthInfo(function(info, done) {
 *       Client.findById(info.clientID, function (err, client) {
 *         info.client = client;
 *         done(err, info);
 *       ***REMOVED***);
 *     ***REMOVED***);
 *
 * @api public
 */
Passport.prototype.transformAuthInfo = function(fn, done) {
  if (typeof fn === 'function') {
    return this._infoTransformers.push(fn);
  ***REMOVED***
  
  // private implementation that traverses the chain of transformers,
  // attempting to transform auth info
  var info = fn;
  
  var stack = this._infoTransformers;
  (function pass(i, err, tinfo) {
    // transformers use 'pass' as an error to skip processing
    if ('pass' === err) {
      err = undefined;
    ***REMOVED***
    // an error or transformed info was obtained, done
    if (err || tinfo) { return done(err, tinfo); ***REMOVED***
    
    var layer = stack[i];
    if (!layer) {
      // if no transformers are registered (or they all pass), the default
      // behavior is to use the un-transformed info as-is
      return done(null, info);
    ***REMOVED***
    
    try {
      var arity = layer.length;
      if (arity == 1) {
        // sync
        var t = layer(info);
        pass(i + 1, null, t);
      ***REMOVED*** else {
        // async
        layer(info, function(e, t) { pass(i + 1, e, t); ***REMOVED*** )
      ***REMOVED***
    ***REMOVED*** catch(e) {
      return done(e);
    ***REMOVED***
  ***REMOVED***)(0);
***REMOVED***

/**
 * Return strategy with given `name`. 
 *
 * @param {String***REMOVED*** name
 * @return {Strategy***REMOVED***
 * @api private
 */
Passport.prototype._strategy = function(name) {
  return this._strategies[name];
***REMOVED***


/**
 * Export default singleton.
 *
 * @api public
 */
exports = module.exports = new Passport();

/**
 * Framework version.
 */
require('pkginfo')(module, 'version');

/**
 * Expose constructors.
 */
exports.Passport = Passport;
exports.Strategy = Strategy;


/**
 * Expose strategies.
 */
exports.strategies = {***REMOVED***;
exports.strategies.SessionStrategy = SessionStrategy;


/**
 * HTTP extensions.
 */
require('./http/request');
