// combined-stream/lib/combined_stream.js
var util = require('util');
var Stream = require('stream').Stream;
var DelayedStream = require('delayed-stream');

module.exports = CombinedStream;
function CombinedStream() {
  this.writable = false;
  this.readable = true;
  this.dataSize = 0;
  this.maxDataSize = 2 * 1024 * 1024;
  this.pauseStreams = true;

  this._released = false;
  this._streams = [];
  this._currentStream = null;
***REMOVED***
util.inherits(CombinedStream, Stream);

CombinedStream.create = function(options) {
  var combinedStream = new this();

  options = options || {***REMOVED***;
  for (var option in options) {
    combinedStream[option] = options[option];
  ***REMOVED***

  return combinedStream;
***REMOVED***;

CombinedStream.isStreamLike = function(stream) {
  return (typeof stream !== 'function')
    && (typeof stream !== 'string')
    && (typeof stream !== 'boolean')    
    && (typeof stream !== 'number')
    && (!Buffer.isBuffer(stream));
***REMOVED***;

CombinedStream.prototype.append = function(stream) {
  var isStreamLike = CombinedStream.isStreamLike(stream);

  if (isStreamLike) {
    if (!(stream instanceof DelayedStream)) {
      stream.on('data', this._checkDataSize.bind(this));

      stream = DelayedStream.create(stream, {
        maxDataSize: Infinity,
        pauseStream: this.pauseStreams,
      ***REMOVED***);
    ***REMOVED***

    this._handleErrors(stream);

    if (this.pauseStreams) {
      stream.pause();
    ***REMOVED***
  ***REMOVED***

  this._streams.push(stream);
  return this;
***REMOVED***;

CombinedStream.prototype.pipe = function(dest, options) {
  Stream.prototype.pipe.call(this, dest, options);
  this.resume();
  return dest;
***REMOVED***;

CombinedStream.prototype._getNext = function() {
  this._currentStream = null;
  var stream = this._streams.shift();


  if (typeof stream == 'undefined') {
    this.end();
    return;
  ***REMOVED***

  if (typeof stream !== 'function') {
    this._pipeNext(stream);
    return;
  ***REMOVED***

  var getStream = stream;
  getStream(function(stream) {
    var isStreamLike = CombinedStream.isStreamLike(stream);
    if (isStreamLike) {
      stream.on('data', this._checkDataSize.bind(this));
      this._handleErrors(stream);
    ***REMOVED***

    this._pipeNext(stream);
  ***REMOVED***.bind(this));
***REMOVED***;

CombinedStream.prototype._pipeNext = function(stream) {
  this._currentStream = stream;

  var isStreamLike = CombinedStream.isStreamLike(stream);
  if (isStreamLike) {
    stream.on('end', this._getNext.bind(this))
    stream.pipe(this, {end: false***REMOVED***);
    return;
  ***REMOVED***

  var value = stream;
  this.write(value);
  this._getNext();
***REMOVED***;

CombinedStream.prototype._handleErrors = function(stream) {
  var self = this;
  stream.on('error', function(err) {
    self._emitError(err);
  ***REMOVED***);
***REMOVED***;

CombinedStream.prototype.write = function(data) {
  this.emit('data', data);
***REMOVED***;

CombinedStream.prototype.pause = function() {
  if (!this.pauseStreams) {
    return;
  ***REMOVED***

  this.emit('pause');
***REMOVED***;

CombinedStream.prototype.resume = function() {
  if (!this._released) {
    this._released = true;
    this.writable = true;
    this._getNext();
  ***REMOVED***

  this.emit('resume');
***REMOVED***;

CombinedStream.prototype.end = function() {
  this._reset();
  this.emit('end');
***REMOVED***;

CombinedStream.prototype.destroy = function() {
  this._reset();
  this.emit('close');
***REMOVED***;

CombinedStream.prototype._reset = function() {
  this.writable = false;
  this._streams = [];
  this._currentStream = null;
***REMOVED***;

CombinedStream.prototype._checkDataSize = function() {
  this._updateDataSize();
  if (this.dataSize <= this.maxDataSize) {
    return;
  ***REMOVED***

  var message =
    'DelayedStream#maxDataSize of ' + this.maxDataSize + ' bytes exceeded.'
  this._emitError(new Error(message));
***REMOVED***;

CombinedStream.prototype._updateDataSize = function() {
  this.dataSize = 0;

  var self = this;
  this._streams.forEach(function(stream) {
    if (!stream.dataSize) {
      return;
    ***REMOVED***

    self.dataSize += stream.dataSize;
  ***REMOVED***);

  if (this._currentStream && this._currentStream.dataSize) {
    this.dataSize += this._currentStream.dataSize;
  ***REMOVED***
***REMOVED***;

CombinedStream.prototype._emitError = function(err) {
  this._reset();
  this.emit('error', err);
***REMOVED***;
