// request/vendor/cookie/index.js
/*!
 * Tobi - Cookie
 * Copyright(c) 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var url = require('url');

/**
 * Initialize a new `Cookie` with the given cookie `str` and `req`.
 *
 * @param {String***REMOVED*** str
 * @param {IncomingRequest***REMOVED*** req
 * @api private
 */

var Cookie = exports = module.exports = function Cookie(str, req) {
  this.str = str;

  // First key is the name
  this.name = str.substr(0, str.indexOf('='));

  // Map the key/val pairs
  str.split(/ *; */).reduce(function(obj, pair){
    pair = pair.split(/ *= */);
    obj[pair[0].toLowerCase()] = pair[1] || true;
    return obj;
  ***REMOVED***, this);

  // Assign value
  this.value = this[this.name];

  // Expires
  this.expires = this.expires
    ? new Date(this.expires)
    : Infinity;

  // Default or trim path
  this.path = this.path || '/';
***REMOVED***;

/**
 * Return the original cookie string.
 *
 * @return {String***REMOVED***
 * @api public
 */

Cookie.prototype.toString = function(){
  return this.str;
***REMOVED***;
