// mongodb/lib/mongodb/connection/repl_set/repl_set_state.js
/**
 * Interval state object constructor
 *
 * @ignore
 */
var ReplSetState = function ReplSetState () {
  this.errorMessages = [];
  this.secondaries = {***REMOVED***;
  this.addresses = {***REMOVED***;
  this.arbiters = {***REMOVED***;
  this.passives = {***REMOVED***;
  this.members = [];
  this.errors = {***REMOVED***;
  this.setName = null;
  this.master = null;
***REMOVED***

ReplSetState.prototype.hasValidServers = function() {
  var validServers = [];
  if(this.master && this.master.isConnected()) return true;

  if(this.secondaries) {
    var keys = Object.keys(this.secondaries)
    for(var i = 0; i < keys.length; i++) {
      if(this.secondaries[keys[i]].isConnected())
        return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***

ReplSetState.prototype.getAllReadServers = function() {
  var candidate_servers = [];
  for(var name in this.addresses) {
    candidate_servers.push(this.addresses[name]);
  ***REMOVED***

  // Return all possible read candidates
  return candidate_servers;
***REMOVED***

ReplSetState.prototype.addServer = function(server, master) {
  server.name = master.me;

  if(master.ismaster) {
    this.master = server;
    this.addresses[server.name] = server;
  ***REMOVED*** else if(master.secondary) {
    this.secondaries[server.name] = server;
    this.addresses[server.name] = server;
  ***REMOVED*** else if(master.arbiters) {
    this.arbiters[server.name] = server;
    this.addresses[server.name] = server;
  ***REMOVED***
***REMOVED***

ReplSetState.prototype.contains = function(host) {
  return this.addresses[host] != null;
***REMOVED***

ReplSetState.prototype.isPrimary = function(server) {
  return this.master && this.master.name == server.name;
***REMOVED***

ReplSetState.prototype.isSecondary = function(server) {
  return this.secondaries[server.name] != null;
***REMOVED***

exports.ReplSetState = ReplSetState;
